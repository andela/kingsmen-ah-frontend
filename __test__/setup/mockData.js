export const initialState = {
  auth: {
    loading: false,
    user: {},
    profile: {},
    errors: {},
    isAuthenticated: true
  },
  article: {
    loading: false,
    errors: {},
    tags: [],
    articles: [],
    article: {}
  }
};
export const userToken = {};
