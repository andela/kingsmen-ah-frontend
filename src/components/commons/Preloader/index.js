import React from 'react';
import Loader from 'react-loader-spinner';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import './index.scss';


function Preloader(props) {
  const { type, styles, height, width, color } = props;
  const className = classNames({
    'main-preloader center justify-center items-center': type === 'page',
    '': type === 'button'
  });

  return (
    <div className={className}>
      <Loader type={styles} height={height} width={width} color={color} className="text-center mx-auto center" />
    </div>
  );
}

Preloader.propTypes = {
  type: PropTypes.string,
  styles: PropTypes.string,
  height: PropTypes.number,
  width: PropTypes.number,
  color: PropTypes.string,
};

Preloader.defaultProps = {
  type: null,
  styles: null,
  height: null,
  width: null,
  color: null
};

export default Preloader;
