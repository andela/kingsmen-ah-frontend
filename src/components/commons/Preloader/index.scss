$loader-bg: rgb(233, 228, 228);

.main-preloader {
  div {
    left: 50%;
    position: absolute;
    top: 50%;
    transform: translate(-50%, -50%);
  }
  background: $loader-bg;
  bottom: 0;
  height: 100vh;
  left: 0;
  opacity: 0.5;
  position: absolute;
  right: 0;
  top: 0;
}

svg {
  margin: 0 auto;
}

.main-preloader div {
  left: 50%;
  position: absolute;
  top: 50%;
  transform: translate(-50%, -50%);
}
