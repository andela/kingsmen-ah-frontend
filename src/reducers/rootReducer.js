import { combineReducers } from 'redux';
import { persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import auth from './auth';
import article from './articles';

const authPersistConfig = {
  key: 'auth',
  storage,
  blacklist: ['errors']
};

const appReducer = combineReducers({
  auth: persistReducer(authPersistConfig, auth),
  article
});

export default appReducer;
