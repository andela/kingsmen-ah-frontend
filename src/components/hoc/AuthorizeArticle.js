import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';
import Preloader from '../commons/Preloader';

/**
 * @function AuthorizeArticle
 * @param {object} props
 * @return {JSX} - MyComponent|Preloader|Redirect
 */
const AuthorizeArticle = (props) => {
  const {
      MyComponent,
      loading,
      article,
      user,
      location
  } = props;

  return (
    <Fragment>
      {loading && (
        <div className="flex w-full h-screen items-center justify-center bg-gray-300">
          <Preloader
            type="page"
            styles="TailSpin"
            height={50}
            width={50}
            color="blue"
          />
        </div>
      )
    }
      {!loading && article.author.id === user.id && <MyComponent {...props} />}
      {!loading && article.author.id !== user.id && (
        <Redirect
          to={{ pathname: `/article/${article.slug}`, state: { from: location } }}
        />
      )}
    </Fragment>
  );
};

AuthorizeArticle.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      articleId: PropTypes.string.isRequired
    }).isRequired
  }).isRequired,
  location: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.object
  ]).isRequired,
  article: PropTypes.shape({
    slug: PropTypes.string.isRequired,
    author: PropTypes.shape({
      id: PropTypes.string.isRequired
    }).isRequired
  }).isRequired,
  user: PropTypes.shape({
    id: PropTypes.string.isRequired
  }).isRequired,
  loading: PropTypes.bool,
  MyComponent: PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.object
  ]).isRequired
};

AuthorizeArticle.defaultProps = {
  loading: false
}

export default AuthorizeArticle;
