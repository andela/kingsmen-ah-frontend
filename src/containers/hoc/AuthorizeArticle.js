import { connect } from 'react-redux';
import AuthorizeArticle from '@components/hoc/AuthorizeArticle';

/**
 * Authorize Article User
 * @function AuthorizeArticle
 * @param {JSX} MyComponent - Dynamic
 * @return {JSX} ConnectedComponent
 */
export default (MyComponent) => {
    const mapStateToProps = state => ({
        article: state.article.article,
        user: state.auth.user,
        loading: state.auth.loading,
        MyComponent
    });

    return connect(mapStateToProps)(AuthorizeArticle);
};