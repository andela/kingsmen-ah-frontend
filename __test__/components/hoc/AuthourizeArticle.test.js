import React from 'react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { createStore, applyMiddleware } from 'redux';
import { Redirect, MemoryRouter } from 'react-router-dom';
import AuthorizeArticle from '@containers/hoc/Authenticator';
import Preloader from '@components/commons/Preloader';
import rootReducer from '@reducers/rootReducer';
import { initialState } from '../../setup/mockData';

const loadingInitialValues = {
    ...initialState,
    auth: { ...initialState.auth, loading: true }
};

const ownArticle = {
  ...initialState,
  article: {
    ...initialState.article,
    article: {
      title: 'Hello World',
      slug: 'hello-world',
      author: {
        id: 1,
      }
    }
  },
  auth: {
    ...initialState.auth,
    user: {
      ...initialState.auth.user,
      id: 1
    },
    loading: false
  }
};

const readerArticle = {
  ...initialState,
  article: {
    ...initialState.article,
    article: {
      title: 'Hello World',
      slug: 'hello-world',
      author: {
        id: 1,
      }
    }
  },
  auth: {
    ...initialState.auth,
    isAuthenticated: false,
    user: {
      ...initialState.auth.user,
      id: 2
    },
    loading: false
  }
};

const authStore = createStore(rootReducer, initialState, applyMiddleware(thunk));
const loadingStore = createStore(rootReducer, loadingInitialValues, applyMiddleware(thunk));
const ownerStore = createStore(rootReducer, ownArticle, applyMiddleware(thunk));
const readerStore = createStore(rootReducer, readerArticle, applyMiddleware(thunk));
const store = createStore(rootReducer, applyMiddleware(thunk));

const setup = () => {
    const dispatch = jest.fn();

    const props = {
      dispatch
    };

    const MockComponent = () => <div>Hi</div>;
    MockComponent.displayName = 'MockComponent';

    const HOCComponent = AuthorizeArticle(MockComponent);
    const location = {
      pathname: '/article/hello-world',
      state: { from: { pathname: '//article/hello-world/edit' } }
    };

    const component = (
      <Provider store={store}>
        <MemoryRouter>
          <HOCComponent location={location} />
        </MemoryRouter>
      </Provider>
      );

    const wrapper = shallow(component);

    const authComponent = (
      <Provider store={authStore}>
        <MemoryRouter>
          <HOCComponent location={location} />
        </MemoryRouter>
      </Provider>
    );

    const authWrapper = mount(authComponent);

    const loadingComponent = (
      <Provider store={loadingStore}>
        <MemoryRouter>
          <HOCComponent location={location} />
        </MemoryRouter>
      </Provider>
    );

    const ownerComponent = (
      <Provider store={ownerStore}>
        <MemoryRouter>
          <HOCComponent location={location} />
        </MemoryRouter>
      </Provider>
    );

    const readerComponent = (
      <Provider store={readerStore}>
        <MemoryRouter>
          <HOCComponent location={location} />
        </MemoryRouter>
      </Provider>
    );

    const loadingWrapper = mount(loadingComponent);
    const ownerWrapper = mount(ownerComponent);
    const readerWrapper = mount(readerComponent);

    const unAuthComponent = (
      <Provider store={store}>
        <MemoryRouter>
          <HOCComponent location={location} />
        </MemoryRouter>
      </Provider>
    );

    const unAuthWrapper = mount(unAuthComponent);

    return {
      props, unAuthWrapper, authWrapper, ownerWrapper, readerWrapper, loadingWrapper, wrapper, MockComponent
    };
};

describe('HOC: Authorize Article', () => {
    afterAll(() => {
        jest.clearAllMocks();
    });

    it('should render correctly', () => {
        const { wrapper } = setup();

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render unauthorized component: Redirect', () => {
        const { readerWrapper, MockComponent } = setup();

        expect(readerWrapper.find(Redirect).length).toBeTruthy();
        expect(readerWrapper.find(MockComponent).length).toBeFalsy();

        readerWrapper.unmount();
    });

    it('should render authorized component: MockComponent', () => {
        const { ownerWrapper, MockComponent } = setup(); 

        expect(ownerWrapper.find(MockComponent).length).toBeTruthy();
        expect(ownerWrapper.find(Redirect).length).toBeFalsy();

        ownerWrapper.unmount();
    });

    it('should render loading component: Loading', () => {
        const { loadingWrapper, MockComponent } = setup();

        expect(loadingWrapper.find(Preloader).length).toBeTruthy();
        expect(loadingWrapper.find(MockComponent).length).toBeFalsy();

        loadingWrapper.unmount();
    });
});
