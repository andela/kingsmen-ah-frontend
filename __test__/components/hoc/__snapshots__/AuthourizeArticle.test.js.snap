// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HOC: Authorize Article should render correctly 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": [Function],
      },
    }
  }
>
  <MemoryRouter>
    <Connect(Authenticator)
      location={
        Object {
          "pathname": "/article/hello-world",
          "state": Object {
            "from": Object {
              "pathname": "//article/hello-world/edit",
            },
          },
        }
      }
    />
  </MemoryRouter>
</ContextProvider>
`;
